analysis/task_analysis_surface/first_level_analysis_ss.py

```diff
@@ -0,0 +1,365 @@
+        #if mid list consists of 5 elements (i.e. 3 canonical contrasts and 2 math symbols), there is always 1 '+' and 1 '-' (in addition to the implied '+ at the beginning)
+        #in that case, we need to multiply the 2 canonical constrasts after the '+' and implied '+' by 0.5
+        if len(events) == 5:
+            plus_multiplier = 0.5
+        #if mid list consists of 7 elements (i.e. 4 canonical contrasts and 3 math symbols), there is always 1 '+' and 2 '-' (in addition to the implied '+ at the beginning)
+        #in that case, we need to multiply the 4 canonical constrasts by 0.5
+        if len(events) == 7:
+            minus_multiplier = 0.5
+            plus_multiplier = 0.5
+            
+        #do the complex contrast calculation
+        for el in events:
+            #math_sign starts of with '+' as set up above since the first canonical contrast has an implied '+' at the beginning (added to the 0 that we used to initialize the complex contrast)
+            if el in ['+','-']:
+                math_sign = el
+            #if current list element isn't a math sign, then it's a canonical contrast and we add/subtract based on the math sign we set
+            #we also multiply the canonical contrast by 1 (default) or 0.5 (certain mid complex contrasts)
+            else:
+                if math_sign == '+':
+                    complex_contrast = complex_contrast + plus_multiplier*canonical_contrasts[el]
+                if math_sign == '-':
+                    complex_contrast = complex_contrast - minus_multiplier*canonical_contrasts[el]
+        #at the end we store the created complex contrast        
+        final_contrasts[task_contrast] = complex_contrast       
+                
+    #add a nuisance regressor contrast to the final set of contrasts that includes all nuisance regressors
+    #firs set up zeros matrix of correct shape
+    final_contrasts['nuisance_regressors']=np.zeros(design_matrix.shape[1])
+    #this iteratively adds a 1 in place of all the confounds in the zeros matrix
+    for confound in selected_confounds.columns:
+        final_contrasts['nuisance_regressors']=final_contrasts['nuisance_regressors']+canonical_contrasts[confound]
+    
+    #return dictionary of all complex contrasts
+    print('finishing up make_localizer_contrasts')
+    return final_contrasts
+
+
+
+
+def create_dscalar(orig_cifti, contrast_output):
+    
+    contrast_output_dict = {'effect_size':contrast_output.effect_size(), 'effect_variance':contrast_output.effect_variance(), 'stat':contrast_output.stat(), 'p_value':contrast_output.p_value(), 'z_score':contrast_output.z_score()}
+    
+    dscalar_output_dict = {}
+    
+    #save old header info
+    orig_cifti_nimg = nimg.load_img(orig_cifti)
+    time_axis, brain_model_axis = [orig_cifti_nimg.header.get_axis(i) for i in range(orig_cifti_nimg.ndim)]
```

Comment:
does this do what you expect it to do, when ndim is more than 2 ? there may be a cleaner way to do this.

Reply:
added commit that catches incorrect number of axes